### PR TITLE
ports/psoc6: Adding deinit to pin and adc machine classes.

### PR DESCRIPTION
**Pull request by jaenrig-ifx**
By creating this pull request you agree to the terms in CONTRIBUTING.md.
https://github.com/Infineon/.github/blob/master/CONTRIBUTING.md
--- DO NOT DELETE ANYTHING ABOVE THIS LINE ---

CONTRIBUTING.md also tells you what to expect in the PR process.

This PR will change the destination branch after #94 is merged. Please, review the other one first.

Description
* Adding obj deinit and module deinit to machine.Pin.
* Adding obj deinit and module deinit to machine.ADC and machine.ADCBlock. Additional cleanup refactoring to accommodate pin_phy and scope of each friend classes.

### CODE REVIEW

File: ports/psoc6/modules/machine/machine_adcblock.c [hunk 1]
```diff
@@ -23,6 +23,39 @@ const adc_block_channel_pin_map_t adc_block_pin_map[] = {
 /******************************************************************************/
 // MicroPython bindings for machine.ADC
 
+// Private helper function to get channel number for provided pin
+int16_t _get_adc_channel_number_for_pin(uint32_t pin) {
+    for (int i = 0; i < (sizeof(adc_block_pin_map) / sizeof(adc_block_pin_map[0])); i++)
```
Comment (NikhitaR-IFX) [inline ports/psoc6/modules/machine/machine_adcblock.c:28]: adc_block_pin_map has "PIN_P10_0". This needs to be changed no as the definition is no more available or?

**Review by NikhitaR-IFX — APPROVED** [review]
All looks good to me!
Just a concern - if a physical pin deinit is done in respective module (for example let's say ADC), then again deiniting it from main has an additional affect? Currently does it check if module using the phy pin has already deinited and if not only then do from also main,c or is it not needed to consider that?

File: ports/psoc6/modules/machine/machine_adcblock.c (hunk 1 above)
Comment (jaenrig-ifx) [reply ports/psoc6/modules/machine/machine_adcblock.c:28]: Yes, we have to check that according to how we integrate that part :)

File: ports/psoc6/main.c [hunk 2]
```diff
@@ -174,7 +177,9 @@ void mpy_task(void *arg) {
 
     // Deinitialize modules
     machine_deinit();
-    adcblock_deinit();
+    mod_pin_deinit();
+    mod_adc_block_deinit();
+    mod_pin_phy_deinit();
```
[the remark this answers is not in the document]
Comment (jaenrig-ifx) [reply ports/psoc6/main.c:182]: Not really. But pin_phy_deinit() is actually redundant, as the others should have taken care of that.

**Comment by jaenrig-ifx** [comment]
> All looks good to me! Just a concern - if a physical pin deinit is done in respective module (for example let's say ADC), then again deiniting it from main has an additional affect? Currently does it check if module using the phy pin has already deinited and if not only then do from also main,c or is it not needed to consider that?

The module deinit when soft reseting will take care of deiniting every instance that was not deinited explicitly.
If the application calls deinit() function then it won´t do anything, because all those objects are already deallocated. deinitialization from pin_phy point of view is only setting the pin to not allocated. But those are static instances, which are not really constructed/destructed dinamically.

**Pull request merged**